src/libsyntax/ext/expand.rs

```diff
@@ -707,6 +778,28 @@ pub fn expand_crate(mut cx: ExtCtxt,
     return (ret, cx.syntax_env.names);
 }
 
+// Expands crate using supplied MacroExpander - allows for
+// non-standard expansion behaviour (e.g. step-wise).
+pub fn expand_crate_with_expander(expander: &mut MacroExpander,
+                                  user_exts: Vec<NamedSyntaxExtension>,
+                                  mut c: Crate) -> (Crate, HashSet<Name>) {
+    init_cx(expander.cx, user_exts, &c);
+
+    let items = SmallVector::many(c.module.items);
+    expander.load_macros(&items);
+    c.module.items = items.into();
```

Comment:
Could you call `expand_crate_with_expander` in `expand_crate` to avoid duplicating this code?